Add Welcome Fragment with TODO list

vitomedved opened the pull request.

vitomedved committed:
Add Welcome Fragment with TODO list

LukaBabic1 commented on app/src/main/java/com/android/template/injection/fragment/FragmentComponentInjects.java:
```diff
@@ -1,5 +1,11 @@
 package com.android.template.injection.fragment;
 
+import com.android.template.ui.welcome.WelcomeFragment;
```
Ajde pogledam ti merge request danas do kraja dana. Ako ćeš raditi dalje na feature-ima, onda napravi branch iz ovog i nastavi sa radom.

vitomedved replied:
Ma samo polako, jos na ovom imam dosta za napraviti. To sam stavio request cisto da pogledas kad stignes

LukaBabic1 commented on app/src/main/java/com/android/template/ui/RouterImpl.java:
```diff
@@ -42,6 +43,11 @@ public boolean showPageInExternalBrowser(final String url) {
 
+    @Override
+    public void showWelcomeScreen() {
+        fragmentManager.beginTransaction().replace(CONTAINER_ID, WelcomeFragment.newInstance(), WelcomeFragment.TAG).commit();
```
Tvoj je projekt, pa ti nemam kaj prigovarati, ali svakako je preglednije kada wrappaš svaki sljedeći chain-ani poziv:
```
fragmentManager.beginTransaction()
                              .replace(CONTAINER_ID, WelcomeFragment.newInstance()
                              .commit()
                           
```

vitomedved replied:
Da, to mi je u biti bio plan već 3 puta napraviti, nikako da se sjetim postaviti opcije

LukaBabic1 commented on app/src/main/java/com/android/template/ui/main/MainActivity.java:
```diff
@@ -1,96 +1,26 @@
-                                                                                                         .build());
-        }
+        // Might change name of the method. In future, you will either be redirected to welcome fragment or some kind of initial screen.
+        presenter.showInitScreen();
```
Ako ova metoda pokazuje fragment interno, onda:

```
if (savedInstanceState == null) {
    presenter.showInitScreen();
}
```

Objašnjenje: https://guides.codepath.com/android/creating-and-using-fragments#looking-up-a-fragment-instance

LukaBabic1 commented on app/src/main/java/com/android/template/ui/main/MainPresenter.java:
```diff
@@ -9,19 +10,23 @@
+
     @Inject
-    GetExampleUseCase getExampleUseCase;
+    Router router;
```
Imaš Router injectan već u BasePresenteru.

LukaBabic1 commented on app/src/main/java/com/android/template/ui/main/MainPresenter.java:
```diff
@@ -9,19 +10,23 @@
-        addDisposable(getExampleUseCase.execute()
+    public void showInitScreen() {
+        router.showWelcomeScreen();
+        /*addDisposable(getExampleUseCase.execute()
```
MOžeš maknuti ovaj zakomentirani kod.

LukaBabic1 commented on app/src/main/res/layout/welcome_fragment.xml:
```diff
@@ -0,0 +1,51 @@
+<?xml version="1.0" encoding="utf-8"?>
+<RelativeLayout xmlns:android="http://schemas.android.com/apk/res/android"
```
Radije se od početka čim više privikavaj na ConstraintLayout, RelativeLayout je so old :D

LukaBabic1 commented on app/src/main/res/layout/welcome_fragment.xml:
```diff
@@ -0,0 +1,51 @@
+    android:layout_height="match_parent">
+
+    <TextView
+        android:id="@+id/welcome_fragment_text"
```
Meni je uvijek draže okrenuti namig, u ovom slučaju:
`fragment_welcome_text` ali svakako mislim da bi ovaj dio `text` mogao biti deskripti niji

LukaBabic1 commented on app/src/main/res/values/dimens.xml:
```diff
@@ -14,5 +14,14 @@
     <dimen name="common_root_card_view_layout_margin_bottom">@dimen/baseline_distance</dimen>
     <dimen name="common_root_card_view_corner_radius">8dp</dimen>
     <dimen name="common_root_card_view_elevation">12dp</dimen>
+    <dimen name="welcome_fragment_margin_bottom">15dp</dimen>
```
U material designu su obično sve dimenzije i veličine višekratnici od 4.

LukaBabic1 commented on app/src/main/res/values/strings.xml:
```diff
@@ -1,3 +1,6 @@
 <resources>
     <string name="app_name">Android Project Template</string>
+    <string name="welcome_fragment_info_text">Welcome to conference app, enter conference ID or click on QR code icon to scan conferences QR code.</string>
```
E, ovo za stringove... Ima nešto što mislim da nisi imao prilike kod nas učiti. Kaj je ideja.
Recimo da imaš string "Hello world from bla bla", u tom slučaju napraviš string resurs koji ima ključ baš `hello_world_from_bla_bla` i vrijednost toga bude "Hello world from bla bla", znači:

`<string name="hello_world_from_bla_bla">Hello world from bla bla</string>`

E, i onda kad taj string želiš negdje iskoristiti, napraviš:
`<string name="welcome_fragment_info_text">@string/hello_world_from_bla_bla</string>`

Ideja je kada ćeš eventualno prevoditi apliakciju, puno je lakše znati što neki tekst na rumunjskom znači ako je ključ resursa "hello_world_from_bla_bla"

vitomedved committed:
Add enter conference ID and scan QR functionality

vitomedved committed:
Minor changes

vitomedved committed:
Fixing merge request issues

vitomedved committed:
Minor change

vitomedved committed:
Add FireBase support
Renamed setConferenceIdUseCase to doesConferenceExistUseCase
Implemented above useCase

vitomedved committed:
Add Get and SetInitialConferenceIdUseCase with implementations.

LukaBabic1 commented on app/src/main/AndroidManifest.xml:
```diff
@@ -2,29 +2,37 @@
+    <uses-permission android:name="android.permission.CAMERA" />
 
     <application
+        android:name=".application.ExampleApplication"
```
Dosta formatting diff-ova ćeš imati. Radije uzmi onaj asseco rules za AS.

vitomedved replied:
Razmišljao sam o tome, ali ne mogu nikako doći do toga s obzirom da je assecovo. Ne znam ako ima di za skinuti taj formatting rules?

LukaBabic1 commented on app/src/main/java/com/android/template/injection/activity/ActivityComponentInjects.java:
```diff
@@ -8,4 +10,8 @@
 
     void inject(MainPresenter presenter);
+
+    //void inject(HomeActivity activity);
```
Makni kod skroz, bolje nego komentiranje.

LukaBabic1 commented on app/src/main/java/com/android/template/injection/application/module/CryptoModule.java:
```diff
@@ -83,6 +85,12 @@ SymmetricKeyProvider provideSymmetricKeyProvider(final DeviceInformation deviceI
+    @Provides
+    @Singleton
+    CryptoEngine provideCryptoEngine(final SymmetricKeyProvider symmetricKeyProvider) {
+        return new CryptoEngineFactoryImpl(symmetricKeyProvider).createCryptoEngineForKeyAlias("IS_THIS_GOOD?");
```
Oke je, ali zakucaj negdje tu konstantu.

LukaBabic1 commented on app/src/main/java/com/android/template/ui/RouterImpl.java:
```diff
@@ -42,8 +44,22 @@ public boolean showPageInExternalBrowser(final String url) {
     private boolean canIntentBeResolved(final Intent intent) {
-        return !listUtils.isEmpty(activity.getPackageManager().queryIntentActivities(intent, 0));
+        return !listUtils.isEmpty(activity.getPackageManager()
+                                          .queryIntentActivities(intent, 0));
```
Nice

LukaBabic1 commented on app/src/main/java/com/android/template/ui/welcome/WelcomeFragment.kt:
```diff
@@ -0,0 +1,81 @@
+    }
+
+    override fun showInvalidConferenceIdError() {
+        toastUtil.showLongToast("You have inputted invalid conference ID or canceled QR scanning.")
```
Radije u resurse.

LukaBabic1 commented on app/src/main/java/com/android/template/ui/welcome/WelcomeFragment.kt:
```diff
@@ -0,0 +1,81 @@
+    }
+
+    override fun showConferenceDoesNotExistError() {
+        toastUtil.showLongToast("Conference with entered ID does not exist. Please try again.")
```
Isto.

LukaBabic1 commented on app/src/main/java/com/android/template/ui/welcome/WelcomeFragment.kt:
```diff
@@ -0,0 +1,81 @@
+    fun onConferenceIdQrInputClick() {
+        // TODO maybe move this to Router class and intent on activity from there
+        // TODO maybe execute as useCase and then return a result (scanned QR code in string)
+        IntentIntegrator.forSupportFragment(this)
```
Da, radije ovo zavrappati u neki utils.

LukaBabic1 commented on app/src/main/java/com/android/template/ui/welcome/WelcomeFragment.kt:
```diff
@@ -0,0 +1,81 @@
+    }
+
+    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
+        val result: IntentResult? = IntentIntegrator.parseActivityResult(requestCode, resultCode, data)
```
Bilo bi dobro provjeriti i request code prije parsiranja.

LukaBabic1 commented on app/src/main/java/com/android/template/ui/welcome/WelcomeFragment.kt:
```diff
@@ -0,0 +1,81 @@
+        val result: IntentResult? = IntentIntegrator.parseActivityResult(requestCode, resultCode, data)
+
+        // TODO: send some kind of indication that QR scanning was canceled so this::showInvalidConferenceIdError() won't be called
+        presenter.checkIfConferenceExists(result?.contents ?: "")
```
Imaš metodu `StringUtils.itOrDefault()` koja se za ovo brine.

LukaBabic1 commented on app/src/main/java/com/android/template/ui/welcome/WelcomePresenter.kt:
```diff
@@ -0,0 +1,72 @@
+    lateinit var resources: Resources
+
+    override fun checkIfConferenceExists(id: String) {
+        // TODO: Is it okay for user to input UID of conference (every conference is saved in firebase database by it's UID) to set initial conference? Or should I provide him with another ID which will reference certain conference?
```
Oke je služiti se vršnim ID-jem.

LukaBabic1 commented on app/src/main/java/com/android/template/ui/welcome/WelcomePresenter.kt:
```diff
@@ -0,0 +1,72 @@
+        Log.e("WelcomePresenter", "DoesConferenceExistUseCase returned an error: $throwable")
+    }
+
+    private fun executeSetInitialConferenceIdUseCase(id: String) {
```
Možda provjera parametra?

vitomedved replied:
Mislim da nije potrebno s obzirom da prvo pozivam DoesConferenceIdExistUseCase i ako on prođe, pozivam ovaj SetInitialConferenceIdUseCase

vitomedved replied:
Doduse sad vidim da treba provjeravati id kod DoesConferenceExistUseCase posto ne mogu neke znakove slati kao request

LukaBabic1 commented on data/src/main/java/template/android/com/data/repository/ConferenceRepositoryImpl.kt:
```diff
@@ -0,0 +1,25 @@
+class ConferenceRepositoryImpl : ConferenceRepository {
+    override fun fetchDoesConferenceIdExist(id: String): Observable<Boolean> {
+        return Observable.create { emitter ->
+            FirebaseDatabase.getInstance().getReference("conferenceIds").child(id).addListenerForSingleValueEvent(object : ValueEventListener {
```
U konstantu `conferenceIds `

LukaBabic1 commented on data/src/main/java/template/android/com/data/repository/ConferenceRepositoryImpl.kt:
```diff
@@ -0,0 +1,25 @@
+
+class ConferenceRepositoryImpl : ConferenceRepository {
+    override fun fetchDoesConferenceIdExist(id: String): Observable<Boolean> {
+        return Observable.create { emitter ->
```
Radije koristi `Single` jer će ti sigurno doći sam jedan podatak.

LukaBabic1 replied:
Ali također provjeri ima li već neki vanjski lib koji mapira Firebase u RxJava2 pozive

vitomedved replied:
Pronašao sam lib. Jel pametno koristiti Maybe<Boolean> kao povratni tip ili da pretvorim taj Maybe u Observable<Boolean>?

LukaBabic1 commented on domain/src/main/java/template/android/com/domain/delegate/ApplicationStorageDelegate.kt:
```diff
@@ -0,0 +1,7 @@
+interface ApplicationStorageDelegate {
+    fun saveConferenceId(id: String)
+
+    fun getConferenceId(): String
```
možda koristiti `com.annimon.Optional`

LukaBabic1 commented on domain/src/main/java/template/android/com/domain/usecase/conference/DoesConferenceExistUseCaseImpl.kt:
```diff
@@ -0,0 +1,11 @@
+
+class DoesConferenceExistUseCaseImpl (private val conferenceRepository: ConferenceRepository) : DoesConferenceExistUseCase {
+    override fun execute(id: String): Observable<Boolean> {
+        // TODO: connect with firebase. Where to implement firebase? Should it be through repository and then implementation in data/firebase package?
```
Da, dobro si to zamislio. Ako će se stvari komplicirati može se kasnijeto još doraditi, al ovo je sasvim dobro!

vitomedved committed:
Fixes, added RxFirebase dependency

vitomedved committed:
Renamed WelcomeFragment to AddConferenceFragment.
Add functionality to skip AddConderenceFragment when conference is already initialized.

vitomedved committed:
Fixing some minor bugs
